clang/tools/pattern-rewriter/src/pattern.cpp

```diff
@@ -0,0 +1,497 @@
+
+  while (PatternIndex < PatternSize && Index < Size) {
+    const auto &Element = Pattern[PatternIndex];
+
+    if (std::holds_alternative<SpacingElement>(Element)) {
+      if (!isWhitespace(Input[Index])) {
+        return {};
+      }
+      while (Index < Size && isWhitespace(Input[Index])) {
+        Index++;
+      }
+      PatternIndex++;
+      continue;
+    }
+
+    if (std::holds_alternative<LiteralElement>(Element)) {
+      const auto &Literal = std::get<LiteralElement>(Element);
+      if (Input[Index] != Literal.Value) {
+        return {};
+      }
+      Index++;
+      PatternIndex++;
+      continue;
+    }
+
+    if (std::holds_alternative<CodeElement>(Element)) {
+      const auto &Code = std::get<CodeElement>(Element);
+      MatchPattern Suffix(Pattern.begin() + PatternIndex + 1,
+                          Pattern.begin() + PatternIndex + 1 +
+                              Code.SuffixLength);
+
+      int Next = parseCodeElement(Suffix, Input, Index);
+      if (Next == -1) {
+        return {};
+      }
+      const int Indentation = detectIndentation(Input, Index);
+      std::string ElementContents =
+          dedent(Input.substr(Index, Next - Index), Indentation);
+      if (Result.Bindings.count(Code.Name)) {
+        if (Result.Bindings[Code.Name] != ElementContents) {
+          return {};
+        }
+      } else {
+        Result.Bindings[Code.Name] = std::move(ElementContents);
+      }
+      Index = Next;
+      PatternIndex++;
+      continue;
+    }
+
+    PatternIndex++;
+  }
+
+  Result.Start = Start;
+  Result.End = Index;
+  return Result;
+}
+
+static void instantiateTemplate(
+    const std::string &Template,
+    const std::unordered_map<std::string, std::string> &Bindings,
+    const int Indentation, std::ostream &OutputStream) {
+  const int Size = Template.size();
+  int Index = 0;
+
+  while (Index < Size && isWhitespace(Template[Index])) {
+    Index++;
+  }
+
+  int End = Size - 1;
+  while (End > (Index + 1) && isWhitespace(Template[End])) {
+    End--;
+  }
+
+  while (Index <= End) {
+    const auto Character = Template[Index];
+
+    if (Index < (Size - 1) && Character == '$' && Template[Index + 1] == '{') {
+      const int BindingStart = Index;
+      Index += 2;
+
+      std::stringstream NameStream;
+      while (Index < Size && Template[Index] != '}') {
+        NameStream << Template[Index];
+        Index++;
+      }
+
+      if (Index == Size) {
+        throw std::runtime_error("Invalid rewrite pattern expression");
+      }
+      Index++;
+
+      std::string Name = NameStream.str();
+      if (Bindings.count(Name) > 0) {
+        const int BindingIndentation =
+            detectIndentation(Template, BindingStart) + Indentation;
+        const std::string Contents =
+            indent(Bindings.at(Name), BindingIndentation);
+        OutputStream << Contents;
+      }
+      continue;
+    }
+
+    OutputStream << Character;
+    if (Character == '\n') {
```

Comment:
consider prepare a std::string(" ", Indentation) in the function begining and call osstream.write(const char *, size_t)